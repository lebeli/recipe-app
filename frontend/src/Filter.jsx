import React, { Component, useState } from "react";
import ExpansionPanel from "@material-ui/core/ExpansionPanel";
import ExpansionPanelSummary from "@material-ui/core/ExpansionPanelSummary";
import ExpansionPanelDetails from "@material-ui/core/ExpansionPanelDetails";
import ExpandMoreIcon from "@material-ui/icons/ExpandMore";
import Typography from "@material-ui/core/Typography";
import Box from "@material-ui/core/Box";
import ToggleButtons from "./ToggleButtons";
import Tags from "./Tags";
import "./Filter.scss";

class Filter extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      ingredients: [
        "Gurke",
        "Tomate",
        "Zwiebel",
        "Knoblauch",
        "Ziegenkäse",
        "Gouda",
        "Erdnüsse"
      ],
      breakfast: false,
      lunch: false,
      dinner: false,
      vegetarian: false,
      vegan: false,
      fast: false,
      i_have_time: false,
      chosen_ingredients: []
    };

    this.updateState = this.updateState.bind(this);
  }

  componentDidMount() {
    const options = {
      method: "GET",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json"
      }
    };

    fetch("/api/ingredients", options)
      .then(response => response.json())
      .then(response => {
        var ingredientsArray = [];
        for (let ingredient of response) {
          ingredientsArray.push(ingredient.name);
        }
        this.setState({
          ingredients: ingredientsArray
        });
      })
      .catch(error => {
        this.setState({
          ingredients: ["Keine Zutaten zur Auswahl"]
        });
      });
  }

  shouldComponentUpdate() {
    console.log("it should");
    return true;
  }

  calculateActiveFilters() {
    let active = 0;
    if (this.state.breakfast == true) {
      active = active + 1;
    }
    if (this.state.lunch == true) {
      active = active + 1;
    }
    if (this.state.dinner == true) {
      active = active + 1;
    }
    if (this.state.vegetarian == true) {
      active = active + 1;
    }
    if (this.state.vegan == true) {
      active = active + 1;
    }
    if (this.state.fast == true) {
      active = active + 1;
    }
    if (this.state.i_have_time == true) {
      active = active + 1;
    }

    active = active + this.state.chosen_ingredients.length;
    return <span>{active} Filter aktiv</span>;
  }

  updateState(name, val) {
    this.setState(
      {
        [name]: val
      },
      () => {
        this.props.updateRecipe(
          this.state.breakfast,
          this.state.lunch,
          this.state.dinner,
          this.state.vegetarian,
          this.state.vegan,
          this.state.i_have_time,
          this.state.fast
        );
      }
    );
  }

  render() {
    return (
      <div className="Filter">
        <ExpansionPanel square id="filter-panel">
          <Box justifyContent="center">
            <Box>
              <ExpansionPanelSummary
                expandIcon={<ExpandMoreIcon id="search-expand-icon" />}
                aria-controls="filter-panel-content"
                id="filter-panel-header-summery"
              >
                <Typography component="span" id="filter-header">
                  Suche eingrenzen ({this.calculateActiveFilters()})
                </Typography>
              </ExpansionPanelSummary>
            </Box>
          </Box>
          <Box display="flex" justifyContent="center" m={1} p={1}>
            <Box p={1}>
              <ExpansionPanelDetails>
                <Typography component="span">
                  <ToggleButtons
                    breakfast={this.state.breakfast}
                    lunch={this.state.lunch}
                    dinner={this.state.dinner}
                    vegetarian={this.state.vegetarian}
                    vegan={this.state.vegan}
                    fast={this.state.fast}
                    i_have_time={this.state.i_have_time}
                    chosen_ingredients={this.state.chosen_ingredients}
                    updateState={this.updateState}
                  />
                  <Tags
                    ingredients={this.state.ingredients}
                    chosen_ingredients={this.state.chosen_ingredients}
                    updateState={this.updateState}
                  />
                </Typography>
              </ExpansionPanelDetails>
            </Box>
          </Box>
        </ExpansionPanel>
      </div>
    );
  }
}

export default Filter;
