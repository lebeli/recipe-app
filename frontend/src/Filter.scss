@import "./styles/styles.scss";

.Filter {
  #filter-panel {
    background-color: $secondaryColor;

    .MuiBox-root {
      display: flex;
      flex-wrap: wrap;
      margin: 0;
      padding: 5px;

      #filter-header {
        font-family: $secondaryFont;
        font-size: 15pt;
        color: white;
        text-align: center;
      }

      #search-expand-icon {
        fill: white;
      }

      .MuiBox-root-34 {
        margin: 0;
      }

      .MuiExpansionPanelSummary-root {
        min-height: 20px;

        .MuiExpansionPanelSummary-content {
          margin: 0;
        }
      }

      .MuiExpansionPanelDetails-root {
        padding: 0;
      }
    }
  }
}

.FilterCategory {
  .MuiToggleButton-root {
    border-radius: 24px;
    border: 0px;
    background-color: white;
    height: 40px;
    padding: 0 40px 0 40px;

    .MuiToggleButton-label {
      font-family: $secondaryFont;
      text-transform: none;
      color: black;
    }
  }

  .MuiToggleButton-root:hover {
    background-color: $primaryColor;

    .MuiToggleButton-label {
      color: white;
    }
  }

  .MuiToggleButton-root.Mui-selected {
    background-color: $primaryColor;

    .MuiToggleButton-label {
      color: white;
    }
  }

  .MuiToggleButton-root.Mui-selected:hover {
    background-color: #d24e13;
  }
}

.Tags {
  max-width: 600px;
  .MuiInputBase-root {
    background-color: white;
    border-radius: 24px;
    padding: 0 0 0 5px;
  }

  .MuiAutocomplete-inputRoot[class*="MuiOutlinedInput-root"] {
    padding: 0 10px 0 10px;
  }

  .MuiOutlinedInput-notchedOutline {
    border: 0px;
    border-radius: 24px;
    height: 50px;
  }

  .MuiChip-root {
    background-color: $primaryColor;
    color: white;
    position: relative;

    .MuiSvgIcon-root {
      fill: white;
    }

    .MuiChip-label {
      font-family: $secondaryFont;
    }
  }

  .MuiInputBase-input {
    font-family: $secondaryFont;
  }

  .MuiIconButton-root {
    background-color: transparent;
  }
}
