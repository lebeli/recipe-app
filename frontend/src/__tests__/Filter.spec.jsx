import React from "react";
import { shallow, mount } from "enzyme";
import Filter from "../Filter";
import renderer from "react-test-renderer";
import "regenerator-runtime";

describe("<Filter />", () => {
  beforeEach(function() {
    global.fetch = jest.fn().mockImplementation(() => {
      var p = new Promise((resolve, reject) => {
        resolve({
          ok: true,
          ingredients: [
            "Gurke",
            "Tomate",
            "Zwiebel",
            "Knoblauch",
            "Ziegenkäse",
            "Gouda",
            "Erdnüsse"
          ],
          json: function() {
            return ingredients;
          }
        });
      });
      return p;
    });
  });

  test("should render elements", async () => {
    const wrapper = mount(<Filter />);
    expect(wrapper.find(".ToggleButtons"));
    expect(wrapper.find(".Tags"));
  });

  test("calculateActiveFilters() should return a description of how many filters are active", () => {
    var mockFunction = () => function mock() {};
    const wrapper = shallow(<Filter updateRecipe={mockFunction} />);

    // Given
    expect(
      wrapper.instance().calculateActiveFilters().props.children[0]
    ).toEqual(0);

    // When
    wrapper.instance().updateState("breakfast", true);

    // Then
    expect(
      wrapper.instance().calculateActiveFilters().props.children[0]
    ).toEqual(1);

    // When
    wrapper.instance().updateState("i_have_time", true);

    // Then
    expect(
      wrapper.instance().calculateActiveFilters().props.children[0]
    ).toEqual(2);
  });

  test("updateState() should update different state attribute", () => {
    var mockFunction = () => function mock() {};
    const wrapper = shallow(<Filter updateRecipe={mockFunction} />);

    // Given
    expect(wrapper.state("fast")).toEqual(false);
    expect(wrapper.state("chosen_ingredients")).toEqual([]);

    // When
    wrapper.instance().updateState("fast", true);
    wrapper.instance().updateState("chosen_ingredients", ["Tomate", "Gurke"]);

    // Then
    expect(wrapper.state("fast")).toEqual(true);
    expect(wrapper.state("chosen_ingredients")).toEqual(["Tomate", "Gurke"]);
  });

  test("renders correctly", () => {
    const tree = renderer.create(<Filter />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
