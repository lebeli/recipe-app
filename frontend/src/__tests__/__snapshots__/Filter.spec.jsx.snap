// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filter /> renders correctly 1`] = `
<div
  className="Filter"
>
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiExpansionPanel-root"
    id="filter-panel"
  >
    <div
      className="MuiBox-root MuiBox-root-188"
    >
      <div
        className="MuiBox-root MuiBox-root-189"
      >
        <div
          aria-controls="filter-panel-content"
          aria-disabled={false}
          aria-expanded={false}
          className="MuiButtonBase-root MuiExpansionPanelSummary-root"
          id="filter-panel-header-summery"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="MuiExpansionPanelSummary-content"
          >
            <span
              className="MuiTypography-root MuiTypography-body1"
              id="filter-header"
            >
              Suche eingrenzen (
              <span>
                0
                 Filter aktiv
              </span>
              )
            </span>
          </div>
          <div
            aria-disabled={false}
            aria-hidden={true}
            className="MuiButtonBase-root MuiIconButton-root MuiExpansionPanelSummary-expandIcon MuiIconButton-edgeEnd"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role={null}
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root"
                focusable="false"
                id="search-expand-icon"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiCollapse-container MuiCollapse-hidden"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              className="MuiBox-root MuiBox-root-190"
            >
              <div
                className="MuiBox-root MuiBox-root-191"
              >
                <div
                  className="MuiExpansionPanelDetails-root"
                >
                  <span
                    className="MuiTypography-root MuiTypography-body1"
                  >
                    <div
                      className="ToggleButtons"
                    >
                      <div
                        className="MuiBox-root MuiBox-root-192"
                      >
                        <div
                          className="MuiBox-root MuiBox-root-193"
                        >
                          <div
                            className="FilterCategory"
                          >
                            <button
                              aria-pressed={false}
                              className="MuiButtonBase-root MuiToggleButton-root"
                              disabled={false}
                              onBlur={[Function]}
                              onChange={[Function]}
                              onClick={[Function]}
                              onDragLeave={[Function]}
                              onFocus={[Function]}
                              onKeyDown={[Function]}
                              onKeyUp={[Function]}
                              onMouseDown={[Function]}
                              onMouseLeave={[Function]}
                              onMouseUp={[Function]}
                              onTouchEnd={[Function]}
                              onTouchMove={[Function]}
                              onTouchStart={[Function]}
                              tabIndex={0}
                              type="button"
                              value="breakfast"
                            >
                              <span
                                className="MuiToggleButton-label"
                              >
                                Frühstück
                              </span>
                            </button>
                          </div>
                        </div>
                        <div
                          className="MuiBox-root MuiBox-root-194"
                        >
                          <div
                            className="FilterCategory"
                          >
                            <button
                              aria-pressed={false}
                              className="MuiButtonBase-root MuiToggleButton-root"
                              disabled={false}
                              onBlur={[Function]}
                              onChange={[Function]}
                              onClick={[Function]}
                              onDragLeave={[Function]}
                              onFocus={[Function]}
                              onKeyDown={[Function]}
                              onKeyUp={[Function]}
                              onMouseDown={[Function]}
                              onMouseLeave={[Function]}
                              onMouseUp={[Function]}
                              onTouchEnd={[Function]}
                              onTouchMove={[Function]}
                              onTouchStart={[Function]}
                              tabIndex={0}
                              type="button"
                              value="lunch"
                            >
                              <span
                                className="MuiToggleButton-label"
                              >
                                Mittagessen
                              </span>
                            </button>
                          </div>
                        </div>
                        <div
                          className="MuiBox-root MuiBox-root-195"
                        >
                          <div
                            className="FilterCategory"
                          >
                            <button
                              aria-pressed={false}
                              className="MuiButtonBase-root MuiToggleButton-root"
                              disabled={false}
                              onBlur={[Function]}
                              onChange={[Function]}
                              onClick={[Function]}
                              onDragLeave={[Function]}
                              onFocus={[Function]}
                              onKeyDown={[Function]}
                              onKeyUp={[Function]}
                              onMouseDown={[Function]}
                              onMouseLeave={[Function]}
                              onMouseUp={[Function]}
                              onTouchEnd={[Function]}
                              onTouchMove={[Function]}
                              onTouchStart={[Function]}
                              tabIndex={0}
                              type="button"
                              value="dinner"
                            >
                              <span
                                className="MuiToggleButton-label"
                              >
                                Abendessen
                              </span>
                            </button>
                          </div>
                        </div>
                      </div>
                      <div
                        className="MuiBox-root MuiBox-root-196"
                      >
                        <div
                          className="MuiBox-root MuiBox-root-197"
                        >
                          <div
                            className="FilterCategory"
                          >
                            <button
                              aria-pressed={false}
                              className="MuiButtonBase-root MuiToggleButton-root"
                              disabled={false}
                              onBlur={[Function]}
                              onChange={[Function]}
                              onClick={[Function]}
                              onDragLeave={[Function]}
                              onFocus={[Function]}
                              onKeyDown={[Function]}
                              onKeyUp={[Function]}
                              onMouseDown={[Function]}
                              onMouseLeave={[Function]}
                              onMouseUp={[Function]}
                              onTouchEnd={[Function]}
                              onTouchMove={[Function]}
                              onTouchStart={[Function]}
                              tabIndex={0}
                              type="button"
                              value="vegetarian"
                            >
                              <span
                                className="MuiToggleButton-label"
                              >
                                vegetarisch
                              </span>
                            </button>
                          </div>
                        </div>
                        <div
                          className="MuiBox-root MuiBox-root-198"
                        >
                          <div
                            className="FilterCategory"
                          >
                            <button
                              aria-pressed={false}
                              className="MuiButtonBase-root MuiToggleButton-root"
                              disabled={false}
                              onBlur={[Function]}
                              onChange={[Function]}
                              onClick={[Function]}
                              onDragLeave={[Function]}
                              onFocus={[Function]}
                              onKeyDown={[Function]}
                              onKeyUp={[Function]}
                              onMouseDown={[Function]}
                              onMouseLeave={[Function]}
                              onMouseUp={[Function]}
                              onTouchEnd={[Function]}
                              onTouchMove={[Function]}
                              onTouchStart={[Function]}
                              tabIndex={0}
                              type="button"
                              value="vegan"
                            >
                              <span
                                className="MuiToggleButton-label"
                              >
                                vegan
                              </span>
                            </button>
                          </div>
                        </div>
                        <div
                          className="MuiBox-root MuiBox-root-199"
                        >
                          <div
                            className="FilterCategory"
                          >
                            <button
                              aria-pressed={false}
                              className="MuiButtonBase-root MuiToggleButton-root"
                              disabled={false}
                              onBlur={[Function]}
                              onChange={[Function]}
                              onClick={[Function]}
                              onDragLeave={[Function]}
                              onFocus={[Function]}
                              onKeyDown={[Function]}
                              onKeyUp={[Function]}
                              onMouseDown={[Function]}
                              onMouseLeave={[Function]}
                              onMouseUp={[Function]}
                              onTouchEnd={[Function]}
                              onTouchMove={[Function]}
                              onTouchStart={[Function]}
                              tabIndex={0}
                              type="button"
                              value="fast"
                            >
                              <span
                                className="MuiToggleButton-label"
                              >
                                schnell
                              </span>
                            </button>
                          </div>
                        </div>
                        <div
                          className="MuiBox-root MuiBox-root-200"
                        >
                          <div
                            className="FilterCategory"
                          >
                            <button
                              aria-pressed={false}
                              className="MuiButtonBase-root MuiToggleButton-root"
                              disabled={false}
                              onBlur={[Function]}
                              onChange={[Function]}
                              onClick={[Function]}
                              onDragLeave={[Function]}
                              onFocus={[Function]}
                              onKeyDown={[Function]}
                              onKeyUp={[Function]}
                              onMouseDown={[Function]}
                              onMouseLeave={[Function]}
                              onMouseUp={[Function]}
                              onTouchEnd={[Function]}
                              onTouchMove={[Function]}
                              onTouchStart={[Function]}
                              tabIndex={0}
                              type="button"
                              value="i_have_time"
                            >
                              <span
                                className="MuiToggleButton-label"
                              >
                                ich habe Zeit
                              </span>
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      className="Tags"
                    >
                      <div
                        aria-expanded={false}
                        aria-owns={null}
                        className="MuiAutocomplete-root"
                        onClick={[Function]}
                        onKeyDown={[Function]}
                        onMouseDown={[Function]}
                        role="combobox"
                      >
                        <div
                          className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
                        >
                          <div
                            className="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
                            onClick={[Function]}
                          >
                            <input
                              aria-activedescendant={null}
                              aria-autocomplete="list"
                              aria-controls={null}
                              aria-invalid={false}
                              autoCapitalize="none"
                              autoComplete="off"
                              autoFocus={false}
                              className="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
                              disabled={false}
                              onAnimationStart={[Function]}
                              onBlur={[Function]}
                              onChange={[Function]}
                              onFocus={[Function]}
                              onMouseDown={[Function]}
                              placeholder="Nach Zutaten suchen..."
                              required={false}
                              spellCheck="false"
                              type="text"
                              value=""
                            />
                            <div
                              className="MuiAutocomplete-endAdornment"
                            >
                              <button
                                aria-label="Clear"
                                className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
                                disabled={false}
                                onBlur={[Function]}
                                onClick={[Function]}
                                onDragLeave={[Function]}
                                onFocus={[Function]}
                                onKeyDown={[Function]}
                                onKeyUp={[Function]}
                                onMouseDown={[Function]}
                                onMouseLeave={[Function]}
                                onMouseUp={[Function]}
                                onTouchEnd={[Function]}
                                onTouchMove={[Function]}
                                onTouchStart={[Function]}
                                tabIndex={-1}
                                title="Clear"
                                type="button"
                              >
                                <span
                                  className="MuiIconButton-label"
                                >
                                  <svg
                                    aria-hidden="true"
                                    className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                                    focusable="false"
                                    role="presentation"
                                    viewBox="0 0 24 24"
                                  >
                                    <path
                                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                                    />
                                  </svg>
                                </span>
                              </button>
                              <button
                                aria-label="Open"
                                className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                                disabled={false}
                                onBlur={[Function]}
                                onClick={[Function]}
                                onDragLeave={[Function]}
                                onFocus={[Function]}
                                onKeyDown={[Function]}
                                onKeyUp={[Function]}
                                onMouseDown={[Function]}
                                onMouseLeave={[Function]}
                                onMouseUp={[Function]}
                                onTouchEnd={[Function]}
                                onTouchMove={[Function]}
                                onTouchStart={[Function]}
                                tabIndex={-1}
                                title="Open"
                                type="button"
                              >
                                <span
                                  className="MuiIconButton-label"
                                >
                                  <svg
                                    aria-hidden="true"
                                    className="MuiSvgIcon-root"
                                    focusable="false"
                                    role="presentation"
                                    viewBox="0 0 24 24"
                                  >
                                    <path
                                      d="M7 10l5 5 5-5z"
                                    />
                                  </svg>
                                </span>
                              </button>
                            </div>
                            <fieldset
                              aria-hidden={true}
                              className="PrivateNotchedOutline-root-176 MuiOutlinedInput-notchedOutline"
                              style={
                                Object {
                                  "paddingLeft": 8,
                                }
                              }
                            >
                              <legend
                                className="PrivateNotchedOutline-legend-177"
                                style={
                                  Object {
                                    "width": 0.01,
                                  }
                                }
                              >
                                <span
                                  dangerouslySetInnerHTML={
                                    Object {
                                      "__html": "&#8203;",
                                    }
                                  }
                                />
                              </legend>
                            </fieldset>
                          </div>
                        </div>
                      </div>
                    </div>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
